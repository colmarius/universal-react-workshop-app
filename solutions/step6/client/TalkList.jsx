import React from 'react'
import Talk from './Talk.jsx'
import axios from 'axios'

const TalkList = React.createClass({

  getInitialState: function () {
    console.log("GET_INITIAL_STATE")
    // We mus setup the state we used to render on server
    if (typeof __PRELOADED_STATE__ !== 'undefined') {
      // Client side
      return {
        talks:__PRELOADED_STATE__ // initial state
      }
    } else {
      // Server side
      return {
        talks: this.props.initialState ? this.props.initialState.talks : []
      }
    }
  },

  // Not needed anymore...it's only for comparing
  componentDidMount: function () {
    console.log("COMPONENT_DID_MOUNT")

    if (typeof __PRELOADED_STATE__ === 'undefined') {
      axios.get('/api/talks')
        .then(res => {
          this.setState({talks: res.data})
      })
    }
  },

  handleVote: function (talkId) {
    axios.post('/api/talk/vote', {id: talkId})
      .then( () => axios.get('/api/talks'))
      .then(res => {
        this.setState({talks: res.data})
      })
  },

  render: function () {
    console.log("RENDERING")
    const talks = this.state.talks.map(talk => {
      return (
        <Talk
          key={'talk-' + talk.id} // used by React
          id={talk.id}
          name={talk.name}
          author={talk.author}
          short={talk.short}
          votes={talk.votes}
          onVote={this.handleVote}
        ></Talk>
      )
    })
    return (<div>{talks}</div>)
  }
})

export default TalkList
