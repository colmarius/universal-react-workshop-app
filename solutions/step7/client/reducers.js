// It's a pure function. If we don't use
// `combineReducers` we don't even need redux stuff
// When a reducer doesn't recognize an action,
// it's supposed/idiomatic to return the state given to it
const reducer = (state = [], action) => {
  switch (action.type) {
    case 'TALKS_UPDATED':
      return action.talks // In this case we repalce the whole state
    default:
      return state
  }
}
export default reducer
