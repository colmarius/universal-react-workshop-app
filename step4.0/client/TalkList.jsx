import React from 'react'
import Talk from './Talk.jsx'

const TalkList = React.createClass({
  render: function () {
    return (
        <div className="list-group">
          <Talk/>
        </div>
    )
  }
})

export default TalkList
