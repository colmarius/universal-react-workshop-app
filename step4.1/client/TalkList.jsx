import React from 'react'
import Talk from './Talk.jsx'
import data from './data.json' // data model

const TalkList = React.createClass({
  getInitialState: function () {
    return {
      talks: [] // initial state
    }
  },

  componentDidMount: function () {
    this.setState({ talks: data })
  },

  handleVote: function (talkId) {
    // We "clone" the array. We have to treat this.state as if it were immutable.
    const updatedTalks = this.state.talks.slice()
    const votedTalk = updatedTalks.find(el => el.id === talkId)
    votedTalk.votes++
    this.setState({ talks: updatedTalks })
  },

  render: function () {

    const talks = this.state.talks.map(talk => {
      return (
        <Talk
          id={talk.id}
          key={'talk-' + talk.id} // used by React
          name={talk.name}
          author={talk.author}
          short={talk.short}
          votes={talk.votes}
          onVote={this.handleVote}
        ></Talk>
      )
    })
    return (
      <div className='talks'>
        {talks}
      </div>
    )
  }
})

export default TalkList
