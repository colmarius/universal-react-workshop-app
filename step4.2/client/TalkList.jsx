import React from 'react'
import Talk from 'Talk'
import data from 'data.json' // data model

const TalkList = React.createClass({
  handleVote: function (talkId) {
    console.log('Voting', talkId)
  },

  getInitialState: function () {
    return {
      talks: [] // initial state
    }
  },

  componentDidMount: function () {
    this.setState({ talks: data })
  },

  handleVote: function (talkId) {
    // We "clone" the array. We have to treat this.state as if it were immutable.
    const updatedTalks = this.state.talks.slice()
    const votedTalk = updatedTalks.find(el => el.id === talkId)
    votedTalk.votes++
    this.setState({ talks: updatedTalks })
  },
  
  render: function () {

    const talks = this.state.talks.map(talk => {
      return (
        <Talk
          key={'talk-' + talk.id} // used by React
          id={talk.id}
          name={talk.name}
          author={talk.author}
          short={talk.short}
          votes={talk.votes}
          onVote={this.handleVote}
        ></Talk>
      )
    })
    return (
      <div className='talks'>
        {talks}
      </div>
    )
  }
})

export default TalkList
