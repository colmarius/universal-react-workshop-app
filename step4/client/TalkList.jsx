import React from 'react'
import Talk from 'Talk'

const TalkList = React.createClass({
  render: function () {
    return (
        <div className="list-group">
          <Talk/>
        </div>
    )
  }
})

export default TalkList
