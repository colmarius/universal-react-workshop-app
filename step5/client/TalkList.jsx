import React from 'react'
import Talk from './Talk.jsx'
import axios from 'axios'

const TalkList = React.createClass({
  getInitialState: function () {
    return {
      talks: []
    }
  },

  componentDidMount: function () {
    axios.get('/api/talks')
      .then((res) => {
        this.setState({ talks: res.data })
      })
  },

  handleVote: function (talkId) {
    axios.post('/api/talk/vote', { id: talkId })
      .then(() => axios.get('/api/talks'))
      .then(res => {
        this.setState({
          talks: res.data
        })
      })
  },

  render: function () {
    const unsortedTalks = this.state.talks.slice()
    const sortedTalks = unsortedTalks.sort((a, b) => b.votes - a.votes)
    const talks = sortedTalks.map(talk => {
      return (
        <Talk
          key={'talk-' + talk.id} // used by React
          id={talk.id}
          name={talk.name}
          author={talk.author}
          short={talk.short}
          votes={talk.votes}
          onVote={this.handleVote}
        ></Talk>
      )
    })
    return (
      <div className='talks'>
        {talks}
      </div>
    )
  }
})

export default TalkList
