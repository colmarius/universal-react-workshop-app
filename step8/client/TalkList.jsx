import React from 'react'
import Talk from './Talk.jsx'
import {connect} from 'react-redux'
import {loadData, vote} from './actions'

const TalkList = React.createClass({

  render: function () { // We should not access state directly anymore!

    const talks = this.props.talks.map(talk => {
      return (
        <Talk
          key={'talk-' + talk.id} // used by React
          id={talk.id}
          name={talk.name}
          author={talk.author}
          short={talk.short}
          votes={talk.votes}
          onVote={this.props.handleVote} // changed, we get the function from props
        ></Talk>
      )
    })
    return (<div>{talks}</div>)
  }
})

const mapStateToProps = (state) => {
  return {
    talks: state
  }
}

const mapDispatchToProps = (dispatch) => {
  return {
    handleVote: (id) => {
      dispatch(vote(id))
    }
  }
}

// Connect Redux state to props and handlers
const TalkListRedux = connect(mapStateToProps, mapDispatchToProps)(TalkList)

export default TalkListRedux
