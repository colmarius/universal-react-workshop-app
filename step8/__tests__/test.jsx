
import React from 'react'
import TalkList from '../client/TalkList'
import {Provider} from 'react-redux'
import reducers from '../client/reducers'
import renderer from 'react-test-renderer'
import {createStore} from 'redux'
const talks = require('../server/data')

test('TalkList test', () => {
  const store = createStore(reducers, talks)
  const component = renderer.create(
    <Provider store={store}>
       <TalkList/>
    </Provider>)
  let talkList = component.toJSON();
  expect(component).toMatchSnapshot();

})
