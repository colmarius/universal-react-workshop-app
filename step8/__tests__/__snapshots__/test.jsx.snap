exports[`test TalkList test 1`] = `
<div>
  <div
    className="list-group-item">
    <a
      onClick={[Function]}
      style={Object {}}>
      <h3
        className="list-group-item-heading">
        Perfect React
      </h3>
    </a>
    <p
      className="list-group-item-text">
      learn perfect react
    </p>
    <h4
      className="list-group-item-text">
      by 
      Marco Piraccini
    </h4>
    <h4
      className="list-group-item-text">
      Number of votes:  
      <span
        className="badge">
        0
      </span>
       
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button">
        Vote
      </button>
    </h4>
  </div>
  <div
    className="list-group-item">
    <a
      onClick={[Function]}
      style={Object {}}>
      <h3
        className="list-group-item-heading">
        Docker in Practice
      </h3>
    </a>
    <p
      className="list-group-item-text">
      become a Docker master
    </p>
    <h4
      className="list-group-item-text">
      by 
      Docker Master
    </h4>
    <h4
      className="list-group-item-text">
      Number of votes:  
      <span
        className="badge">
        0
      </span>
       
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button">
        Vote
      </button>
    </h4>
  </div>
  <div
    className="list-group-item">
    <a
      onClick={[Function]}
      style={Object {}}>
      <h3
        className="list-group-item-heading">
        Arduino Everywhere
      </h3>
    </a>
    <p
      className="list-group-item-text">
      learn arduino
    </p>
    <h4
      className="list-group-item-text">
      by 
      Archimede Pitagorico
    </h4>
    <h4
      className="list-group-item-text">
      Number of votes:  
      <span
        className="badge">
        0
      </span>
       
      <button
        className="btn btn-default"
        disabled={false}
        onClick={[Function]}
        type="button">
        Vote
      </button>
    </h4>
  </div>
</div>
`;
